PhysicsTools/NanoAOD/python/taus_cff.py

```diff
@@ -31,22 +35,44 @@
 
 ##################### Tables for final output and docs ##########################
 def _tauIdWPMask(pattern, choices, doc=""):
-    return Var(" + ".join(["%d * tauID('%s')" % (pow(2,i), pattern % c) for (i,c) in enumerate(choices)]), "uint8", 
-               doc=doc+": bitmask "+", ".join(["%d = %s" % (pow(2,i),c) for (i,c) in enumerate(choices)]))
+    return Var(" + ".join(["tauID('%s')" % (pattern % c) for c in choices]), "uint8", 
+               doc=doc+": "+", ".join(["%d = %s" % (i,c) for (i,c) in enumerate(choices, start=1)]))
+def _tauIdWPMask_from_raw(pattern, wp_thrs, choices, doc=""):
+    assert wp_thrs is not None, "wp_thrs argument in _tauIdWPMask_from_raw() is None, expect it to be dict-like"
+    vs_type = pattern.split('VS')[-1]
+    if vs_type not in ['e', 'mu', 'jet']:
+        raise ValueError("in WP from raw computation, vs_type should be in [e,mu,jet], got %s" % vs_type)
+    
+    var_definition = []
+    for wp_name in choices:
+        if not isinstance(wp_thrs[vs_type][wp_name], float):
+            raise TypeError("Threshold for vs_type=%s and WP=%s is not a float number." % (vs_type, wp_name))
+        wp_definition = "(tauID('%s')>%.4f)" % (pattern % wp_name, wp_thrs[vs_type][wp_name])
+        var_definition.append(wp_definition)
+    var_definition = " + ".join(var_definition)
+
+    return Var(var_definition, "uint8", doc=doc+": "+", ".join(["%d = %s" % (i,wp_name) for (i,wp_name) in enumerate(choices, start=1)]))
+
 def _tauId2WPMask(pattern,doc):
     return _tauIdWPMask(pattern,choices=("Loose","Tight"),doc=doc)
 def _tauId3WPMask(pattern,doc):
     return _tauIdWPMask(pattern,choices=("Loose","Medium","Tight"),doc=doc)
-def _tauId4WPMask(pattern,doc):
-    return _tauIdWPMask(pattern, choices=("VLoose", "Loose", "Medium", "Tight"), doc=doc)
+def _tauId4WPMask(pattern,doc,from_raw=False,wp_thrs=None):
+    if from_raw:
+        return _tauIdWPMask_from_raw(pattern, wp_thrs, choices=("VLoose", "Loose", "Medium", "Tight"), doc=doc)
+    else:
+        return _tauIdWPMask(pattern,choices=("VLoose", "Loose", "Medium", "Tight"),doc=doc)
 def _tauId5WPMask(pattern,doc):
```

Comment:
can we get rid of `_tauId*` functions for DeepTau v2p1 and 2p5? I.e we can pass WP dict in both cases, the only thing we need to worry is the order, but it can be done with custom sort (or, better by having ordered dict in the first place)

Reply:
Yes, that makes sense to remove code duplication, will update it to have only one single `_tauIdWPMask` function. 

Regarding the ordered list, I think we actually don't need it, since it is sufficient to pass a list (or tuple) of WP names as `choices`, which is an already ordered object. Or did you mean to enforce internally inside of the function the custom order (opposite to having the user to ensure it is correct)?

Reply:
done: [141182c](https://github.com/ofivite/cmssw/pull/2/commits/141182cfcb235985b4a41e05927491e29959d8c3)